### PR TITLE
Improve password tool api features

### PR DESCRIPTION
**Pull request by DFolchA**
|Related issue|
|---|
|#1638|

## Description

 This PR improves how the password tool changes the API passwords and applies the changes to the installation assistant.

### CODE REVIEW

File: unattended_installer/passwords_tool/passwordsMain.sh [hunk 1]
```diff
@@ -240,37 +221,39 @@ function main() {
...
 
-        if [ -n "${changeall}" ]; then
+        if [ -n "${changeall}" ] || [ -n "${p_file}" ]; then
             passwords_readUsers
-            passwords_generatePassword
+            if [ -n "${adminUser}" ] && [ -n "${adminPassword}" ]; then
+                passwords_getApiToken
+                passwords_getApiUsers
+                passwords_getApiIds
+            else
```
Comment (okynos) [inline unattended_installer/passwords_tool/passwordsMain.sh:230]: Death code, else never visited.

**Review by okynos — CHANGES_REQUESTED** [review]

File: unattended_installer/passwords_tool/passwordsMain.sh (hunk 1 above)
Comment (DFolchA) [reply unattended_installer/passwords_tool/passwordsMain.sh:230]: The code in the else clause is executed when the admin passwords are not provided.

File: unattended_installer/install_functions/checks.sh [hunk 2]
```diff
@@ -135,23 +135,14 @@ function checks_arguments() {
     # -------------- Wazuh ------------------------------------------
 
     if [ -n "${wazuh}" ]; then
-        if [ -n "${wazuh_installed}" ] || [ -n "${wazuh_remaining_files}" ]; then
+        if [ -n "${wazuh_installed}" ] || [ -n "${wazuh_remaining_files}" ] || [ -n "${filebeat_installed}" ] || [ -n "${filebeat_remaining_files}" ]; then
             if [ -n "${overwrite}" ]; then
                 installCommon_rollBack
             else
                 common_logger -e "Wazuh is already installed in this node or some of its files haven't been erased. Use option -o|--overwrite to overwrite all components."
```
Comment (alberpilot) [inline unattended_installer/install_functions/checks.sh:142]: What about if a user receives this message and wants to manually remove the component? it's not possible with the information presented.

File: unattended_installer/install_functions/installCommon.sh [hunk 3]
```diff
@@ -168,6 +201,15 @@ function installCommon_changePasswords() {
             changeall=1
             passwords_readUsers
         fi
+        set -x
```
Comment (alberpilot) [inline unattended_installer/install_functions/installCommon.sh:204]: Remove that

File: unattended_installer/passwords_tool/passwordsFunctions.sh [hunk 4]
```diff
@@ -172,19 +224,70 @@ function passwords_generatePasswordFile() {
...
+        echo "# ${api_user_description[${i}]}" >> "${gen_file}"
+        echo "  api_username: '${api_users[${i}]}'" >> "${gen_file}"
+        echo "  api_password: '${api_passwords[${i}]}'" >> "${gen_file}"
         echo ""	>> "${gen_file}"
     done
 
 }
 
+function passwords_getApiToken() {
+    TOKEN_API=$(curl -s -u "${adminUser}":"${adminPassword}" -k -X GET "https://localhost:55000/security/user/authenticate?raw=true")
```
Comment (alberpilot) [inline unattended_installer/passwords_tool/passwordsFunctions.sh:251]: Empty line after function declaration, please

Comment (alberpilot) [reply unattended_installer/passwords_tool/passwordsFunctions.sh:251]: Same for all others.

**Review by alberpilot — CHANGES_REQUESTED** [review]
Please review the requested changes

File: unattended_installer/install_functions/checks.sh (hunk 2 above)
Comment (DFolchA) [reply unattended_installer/install_functions/checks.sh]: Done [9822e92](https://github.com/wazuh/wazuh-packages/pull/1663/commits/9822e923dc31c00b848cf303afa79abc58707ea5)

**Comment by rauldpm** [comment]
I have tried to change the API passwords in the following way:

```
[root@wazuh-server wazuh-user]# bash wazuh-passwords-tool.sh -f pass.yml 
20/06/2022 18:03:42 INFO: Wazuh API admin credentials not provided, Wazuh API passwords not changed
20/06/2022 18:03:58 INFO: The password for user admin is admin
20/06/2022 18:03:58 INFO: The password for user kibanaserver is kibanaserver
20/06/2022 18:03:58 INFO: The password for user kibanaro is kibanaro
20/06/2022 18:03:58 INFO: The password for user logstash is logstash
20/06/2022 18:03:58 INFO: The password for user readall is readall
20/06/2022 18:03:58 INFO: The password for user snapshotrestore is snapshotrestore
20/06/2022 18:03:58 INFO: The password for user wazuh_admin is wazuh_admin
20/06/2022 18:03:58 INFO: The password for user wazuh_user is wazuh_user
20/06/2022 18:03:58 WARNING: Wazuh indexer passwords changed. Remember to update the password in the Wazuh dashboard and Filebeat nodes if necessary, and restart the services.
[root@wazuh-server wazuh-user]# bash wazuh-passwords-tool.sh -f pass.yml -au admin -ap admin
20/06/2022 18:04:09 ERROR: Invalid admin user credentials
[root@wazuh-server wazuh-user]# bash wazuh-passwords-tool.sh -f pass.yml -au wazuh -ap wazuh
20/06/2022 18:04:28 INFO: The password for user admin is admin
20/06/2022 18:04:28 INFO: The password for user kibanaserver is kibanaserver
20/06/2022 18:04:28 INFO: The password for user kibanaro is kibanaro
20/06/2022 18:04:28 INFO: The password for user logstash is logstash
20/06/2022 18:04:28 INFO: The password for user readall is readall
20/06/2022 18:04:28 INFO: The password for user snapshotrestore is snapshotrestore
20/06/2022 18:04:28 INFO: The password for user wazuh_admin is wazuh_admin
20/06/2022 18:04:28 INFO: The password for user wazuh_user is wazuh_user
20/06/2022 18:04:28 WARNING: Wazuh indexer passwords changed. Remember to update the password in the Wazuh dashboard and Filebeat nodes if necessary, and restart the services.
20/06/2022 18:04:28 INFO: The password for Wazuh API user wazuh is wazuh1
20/06/2022 18:04:29 INFO: The password for Wazuh API user wazuh-wui is wazuh-wui1
20/06/2022 18:04:29 INFO: Updated wazuh-wui user password in wazuh dashboard. Remember to restart the service.
[root@wazuh-server wazuh-user]#
```

I have used the credentials `wazuh:wazuh` to change the password of the user `wazuh` to `wazuh1`, after this change, the WUI is not accessible:

image

```
INFO: Current API id [default]
INFO: Checking current API id [default]...
INFO: Current API id [default] has some problem: 3002 - Request failed with status code 403
INFO: Getting API hosts...
INFO: API hosts found: 1
INFO: Checking API host id [default]...
INFO: Could not connect to API id [default]: 3099 - ERROR3099 - Limit of login attempts reached. The current IP has been blocked due to a high number of login attempts
INFO: Removed [navigate] cookie
ERROR: No API available to connect
```

When trying to restore the password by changing the password of the wazuh user from `wazuh1` to `wazuh`, the script indicates that it cannot find the api users

```
[root@wazuh-server wazuh-user]# bash wazuh-passwords-tool.sh -f pass.yml -au wazuh -ap wazuh1
20/06/2022 18:09:22 ERROR: The Wazuh API user wazuh does not exist
20/06/2022 18:09:22 ERROR: The Wazuh API user wazuh-wui does not exist
20/06/2022 18:09:34 INFO: The password for user admin is admin
20/06/2022 18:09:34 INFO: The password for user kibanaserver is kibanaserver
20/06/2022 18:09:34 INFO: The password for user kibanaro is kibanaro
20/06/2022 18:09:34 INFO: The password for user logstash is logstash
20/06/2022 18:09:34 INFO: The password for user readall is readall
20/06/2022 18:09:34 INFO: The password for user snapshotrestore is snapshotrestore
20/06/2022 18:09:34 INFO: The password for user wazuh_admin is wazuh_admin
20/06/2022 18:09:34 INFO: The password for user wazuh_user is wazuh_user
20/06/2022 18:09:34 WARNING: Wazuh indexer passwords changed. Remember to update the password in the Wazuh dashboard and Filebeat nodes if necessary, and restart the services.
```

File: unattended_installer/passwords_tool/passwordsMain.sh [hunk 5]
```diff
@@ -86,16 +84,10 @@ function main() {
...
             "-A"|"--api")
                 api=1
-                if [ -z ${2} ]; then
-                    echo "Argument -A|--api needs a second argument"
-                    getHelp
-                    exit 1
-                fi
                 currentPassword=${2}
-                shift 2
+                shift 1
```
Comment (okynos) [inline unattended_installer/passwords_tool/passwordsMain.sh:88]: Does it needs `Shift 2`? or include the $2 check?

**Review by okynos — CHANGES_REQUESTED** [review]
Minor change, GJ.

File: unattended_installer/passwords_tool/passwordsMain.sh (hunk 5 above)
Comment (DFolchA) [reply unattended_installer/passwords_tool/passwordsMain.sh:87]: Done [8455a7f](https://github.com/wazuh/wazuh-packages/pull/1663/commits/8455a7f193cc2dccb41818cd5e4d406fa7018f27)

**Comment by rauldpm** [comment]
Hello @DFolchA 

I am testing the wazuh-passwords-tool.sh script and i see this output, please take a look at

```
wazuh-passwords-tool.sh: line 347: [ wazuh: command not found
wazuh-passwords-tool.sh: line 347: [ wazuh-wui: command not found
```

<details><summary>Change password output command not found</summary>

```
[root@centos7 unattended_installer]# bash wazuh-passwords-tool.sh -f wazuh-install-files/passwords.wazuh -au wazuh -ap A2fn8*JOE+8gCGSN2tg?*Rma.k3?4cwF
23/06/2022 16:19:37 INFO: The password for user admin is p+CU+GDzSc3YF*m9EH8jRYJccjItkJ8W
23/06/2022 16:19:37 INFO: The password for user kibanaserver is xsJkdx7PMClHCpDfQ2NC1KeQnv1K*E?c
23/06/2022 16:19:37 INFO: The password for user kibanaro is oCwu*jG341LccSND7Mt7N64ug7QASTlO
23/06/2022 16:19:37 INFO: The password for user logstash is .g0nj8pyMvqBLzuHFeeKrCDCtf92h?fX
23/06/2022 16:19:37 INFO: The password for user readall is XRcaQsM*TvjQ*rM*35MMHjHoJjL?eLXP
23/06/2022 16:19:37 INFO: The password for user snapshotrestore is xPLXIfSq0QLdU9sMA*D16WE3jivj+o4P
23/06/2022 16:19:37 INFO: The password for user wazuh_admin is D**BqZqOR+GbN*PRvvTDPH8e535LSeyj
23/06/2022 16:19:37 INFO: The password for user wazuh_user is mVz.lXimR7vI2w3X.kyPiu3P*4nY0+Y.
23/06/2022 16:19:37 WARNING: Wazuh indexer passwords changed. Remember to update the password in the Wazuh dashboard and Filebeat nodes if necessary, and restart the services.
wazuh-passwords-tool.sh: line 347: [ wazuh: command not found
23/06/2022 16:19:37 INFO: The password for Wazuh API user wazuh is A2fn8*JOE+8gCGSN2tg?*Rma.k3?4cwF
wazuh-passwords-tool.sh: line 347: [ wazuh-wui: command not found
23/06/2022 16:19:37 INFO: The password for Wazuh API user wazuh-wui is 9rgH8d*34GANwb3sIuDRh5+oaAbLgoHV
23/06/2022 16:19:37 INFO: Updated wazuh-wui user password in wazuh dashboard. Remember to restart the service.
[root@centos7 unattended_installer]#
```

</details>

**Review by alberpilot — CHANGES_REQUESTED** [review]
Please check the following result:

```
[root@ip-172-31-24-117 unattended_installer]# bash wazuh-install.sh -wd dashboard
23/06/2022 16:35:31 INFO: Starting Wazuh installation assistant. Wazuh version: 4.3.4
23/06/2022 16:35:31 INFO: Verbose logging redirected to /var/log/wazuh-install.log
23/06/2022 16:35:36 INFO: Wazuh repository added.
23/06/2022 16:35:36 INFO: --- Wazuh dashboard ----
23/06/2022 16:35:36 INFO: Starting Wazuh dashboard installation.
23/06/2022 16:36:20 INFO: Wazuh dashboard installation finished.
23/06/2022 16:36:20 INFO: Wazuh dashboard post-install configuration finished.
23/06/2022 16:36:20 INFO: Starting service wazuh-dashboard.
23/06/2022 16:36:20 INFO: wazuh-dashboard service started.
23/06/2022 16:36:33 INFO: Updated wazuh-wui user password in wazuh dashboard. Remember to restart the service.
23/06/2022 16:36:38 INFO: Initializing Wazuh dashboard web application.
23/06/2022 16:36:38 INFO: Wazuh dashboard web application initialized.
23/06/2022 16:36:38 INFO: --- Summary ---
23/06/2022 16:36:38 INFO: You can access the web interface https://172.31.24.117
    User: admin
    Password: Ysdasg34g423234r=-=_e22l55
23/06/2022 16:36:38 INFO: Installation finished.
```
The message:
```
23/06/2022 16:36:33 INFO: Updated wazuh-wui user password in wazuh dashboard. Remember to restart the service.
```
is not correct in an unattended run.

**Review by rauldpm — CHANGES_REQUESTED** [review]
I am testing the wazuh-passwords-tool.sh script and i see this output, please take a look at

```
wazuh-passwords-tool.sh: line 347: [ wazuh: command not found
wazuh-passwords-tool.sh: line 347: [ wazuh-wui: command not found
```

<details><summary>Change password output command not found</summary>

```
[root@centos7 unattended_installer]# bash wazuh-passwords-tool.sh -f wazuh-install-files/passwords.wazuh -au wazuh -ap A2fn8*JOE+8gCGSN2tg?*Rma.k3?4cwF
23/06/2022 16:19:37 INFO: The password for user admin is p+CU+GDzSc3YF*m9EH8jRYJccjItkJ8W
23/06/2022 16:19:37 INFO: The password for user kibanaserver is xsJkdx7PMClHCpDfQ2NC1KeQnv1K*E?c
23/06/2022 16:19:37 INFO: The password for user kibanaro is oCwu*jG341LccSND7Mt7N64ug7QASTlO
23/06/2022 16:19:37 INFO: The password for user logstash is .g0nj8pyMvqBLzuHFeeKrCDCtf92h?fX
23/06/2022 16:19:37 INFO: The password for user readall is XRcaQsM*TvjQ*rM*35MMHjHoJjL?eLXP
23/06/2022 16:19:37 INFO: The password for user snapshotrestore is xPLXIfSq0QLdU9sMA*D16WE3jivj+o4P
23/06/2022 16:19:37 INFO: The password for user wazuh_admin is D**BqZqOR+GbN*PRvvTDPH8e535LSeyj
23/06/2022 16:19:37 INFO: The password for user wazuh_user is mVz.lXimR7vI2w3X.kyPiu3P*4nY0+Y.
23/06/2022 16:19:37 WARNING: Wazuh indexer passwords changed. Remember to update the password in the Wazuh dashboard and Filebeat nodes if necessary, and restart the services.
wazuh-passwords-tool.sh: line 347: [ wazuh: command not found
23/06/2022 16:19:37 INFO: The password for Wazuh API user wazuh is A2fn8*JOE+8gCGSN2tg?*Rma.k3?4cwF
wazuh-passwords-tool.sh: line 347: [ wazuh-wui: command not found
23/06/2022 16:19:37 INFO: The password for Wazuh API user wazuh-wui is 9rgH8d*34GANwb3sIuDRh5+oaAbLgoHV
23/06/2022 16:19:37 INFO: Updated wazuh-wui user password in wazuh dashboard. Remember to restart the service.
[root@centos7 unattended_installer]#
```

</details>

File: unattended_installer/install_functions/installCommon.sh [hunk 6]
```diff
@@ -273,17 +318,19 @@ function installCommon_installPrerequisites() {
...
-  password: password
+  indexer_username: name
+  indexer_password: password
 
-# Wazuh indexer admin user
-  username: kibanaserver
-  password: NiwXQw82pIf0dToiwczduLBnUPEvg7T0
+# Description
+  api_username: kibanaserver
+  api_password: NiwXQw82pIf0dToiwczduLBnUPEvg7T0
```
Comment (rauldpm) [inline unattended_installer/install_functions/installCommon.sh:333]: Why? This is a help message not related to indexer

File: unattended_installer/install_functions/installVariables.sh [hunk 7]
```diff
@@ -55,3 +55,6 @@ readonly dashboard_rpm_base_url="${base_url}/yum"
 readonly dashboard_rpm_package="wazuh-dashboard-${wazuh_version}-${dashboard_revision_rpm}.x86_64.rpm"
 readonly wazuh_gpg_key="https://${bucket}/key/GPG-KEY-WAZUH"
 readonly filebeat_config_file="${resources}/tpl/wazuh/filebeat/filebeat.yml"
+
+adminUser="wazuh"
+adminPassword="wazuh"
```
Comment (rauldpm) [inline unattended_installer/install_functions/installVariables.sh:60]: Is this necessary?

File: unattended_installer/passwords_tool/passwordsFunctions.sh [hunk 8]
```diff
@@ -223,65 +369,112 @@ function passwords_readAdmincerts() {
...
-  password: password
+  indexer_username: name
+  indexer_password: password
 
-# Wazuh indexer admin user
-  username: kibanaserver
-  password: NiwXQw82pIf0dToiwczduLBnUPEvg7T0
+# Description
+  api_username: kibanaserver
+  api_password: NiwXQw82pIf0dToiwczduLBnUPEvg7T0
```
Comment (rauldpm) [inline unattended_installer/passwords_tool/passwordsFunctions.sh:384]: Same, not related to indexer users

File: unattended_installer/passwords_tool/passwordsFunctions.sh [hunk 9]
```diff
@@ -379,56 +572,10 @@ function passwords_runSecurityAdmin() {
...
-
-    if [ -f "/usr/share/wazuh-dashboard/data/wazuh/config/wazuh.yml" ]; then
-        password_wazuh_wui=$(grep -A 1 "username: 'wazuh_wui'" "${p_file}" | tail -n1 | awk -F': ' '{print $2}' | sed -e "s/[\'\"]//g")
-        eval 'sed -i "s|password: wazuh-wui|password: ${password_wazuh_wui}|g" /usr/share/wazuh-dashboard/data/wazuh/config/wazuh.yml'
-    else
-        common_logger -e "File /usr/share/wazuh-dashboard/data/wazuh/config/wazuh.yml does not exist"
-    fi
-
-}
+}
```
Comment (rauldpm) [inline unattended_installer/passwords_tool/passwordsFunctions.sh:581]: Add end of line please

**Review by rauldpm — CHANGES_REQUESTED** [review]
Review please

File: unattended_installer/install_functions/installVariables.sh (hunk 7 above)
Comment (DFolchA) [reply unattended_installer/install_functions/installVariables.sh:60]: This is used to change the API passwords.

File: unattended_installer/install_functions/installCommon.sh (hunk 6 above)
Comment (DFolchA) [reply unattended_installer/install_functions/installCommon.sh:333]: This comment indicates the correct format for the file it should show the two options in the new format.

File: unattended_installer/passwords_tool/passwordsFunctions.sh (hunk 8 above)
Comment (DFolchA) [reply unattended_installer/passwords_tool/passwordsFunctions.sh:384]: This comment indicates the correct format for the file it should show the two options in the new format.

File: unattended_installer/install_functions/installCommon.sh (hunk 6 above)
Comment (DFolchA) [reply unattended_installer/install_functions/installCommon.sh:333]: Messages improved in [8d36492](https://github.com/wazuh/wazuh-packages/pull/1663/commits/8d36492b436811733ca8766e0cd6ec68929fb46f)

Comment (rauldpm) [inline unattended_installer/install_functions/installCommon.sh:333]: Use help format (<user> <password>) without specific values

```
                Wazuh indexer users must have this format

                    # Description
                      indexer_username: <user>
                      indexer_password: <password>
                Wazuh API users must have this format

                    # Description
                      api_username: <user>
                      api_password: <password>
```

**Review by rauldpm — CHANGES_REQUESTED** [review]

File: unattended_installer/passwords_tool/passwordsMain.sh [hunk 10]
```diff
@@ -50,11 +48,16 @@ function getHelp() {
     echo -e ""
     echo -e "        -f,  --file <passwords.wazuh>"
     echo -e "                Changes the passwords for the ones given in the file."
-    echo -e "                Each user has to have this format."
+    echo -e "                Wazuh indexer users must have this format"
```
Comment (rauldpm) [inline unattended_installer/passwords_tool/passwordsMain.sh:51]: Add a separation line here please 

```
-f,  --file <passwords.wazuh>
                Changes the passwords for the ones given in the file.
                Wazuh indexer users must have this format

                    # Description
                      indexer_username: <user>
                      indexer_password: <password>
                Wazuh API users must have this format

                    # Description
                      api_username: <user>
                      api_password: <password>

```

**Review by rauldpm — CHANGES_REQUESTED** [review]

File: unattended_installer/install_functions/installCommon.sh (hunk 6 above)
Comment (DFolchA) [reply unattended_installer/install_functions/installCommon.sh]: Done [d79e4c7](https://github.com/wazuh/wazuh-packages/pull/1663/commits/d79e4c7c62d01fcde35be5936fea2bf0c36c2a59)

**Review by rauldpm — APPROVED** [review]
LGTM

**Review by alberpilot — APPROVED** [review]

**Pull request merged**